crates/cast/src/lib.rs

```diff
@@ -816,6 +820,101 @@ where
+    pub async fn convert_block_number(
+        &self,
+        block: Option<BlockId>,
+    ) -> Result<Option<BlockNumber>, eyre::Error> {
+        match block {
+            Some(block) => match block {
+                BlockId::Number(block_number) => Ok(Some(block_number)),
+                BlockId::Hash(hash) => {
+                    let block = self.provider.get_block(hash).await?;
+                    Ok(block.map(|block| block.number.unwrap()).map(BlockNumber::from))
+                }
+            },
+            None => Ok(None),
+        }
+    }
+
+    pub async fn subscribe(
+        &self,
+        filter: Filter,
+        output: &mut dyn io::Write,
+        to_json: bool,
+    ) -> Result<()>
+    where
+        <M as Middleware>::Provider: PubsubClient,
+    {
+        // Initialize the subscription stream for logs
+        let mut subscription = self.provider.subscribe_logs(&filter).await?;
+
+        // Check if a to_block is specified, if so, subscribe to blocks
+        let mut block_subscription = if filter.get_to_block().is_some() {
+            Some(self.provider.subscribe_blocks().await?)
+        } else {
+            None
+        };
+
+        let to_block_number = filter.get_to_block();
+
+        // If output should be JSON, start with an opening bracket
+        if to_json {
+            write!(output, "[")?;
+        }
+
+        let mut first = true;
+
+        loop {
+            tokio::select! {
+                // If block subscription is present, listen to it to avoid blocking indefinitely past the desired to_block
+                block = if let Some(bs) = &mut block_subscription {
+                    Either::Left(bs.next().fuse())
+                } else {
+                    Either::Right(futures::future::pending())
+                } => {
+                    if let (Some(block), Some(to_block)) = (block, to_block_number) {
+                        if block.number.map_or(false, |bn| bn > to_block) {
+                            break;
+                        }
+                    }
+                },
+                // Process incoming log
+                log = subscription.next() => {
+                    if to_json {
+                        if !first {
+                            write!(output, ",")?;
+                        }
+                        first = false;
+                        let log_str = serde_json::to_string(&log).unwrap();
+                        write!(output, "{}", log_str)?;
+                    } else {
+                        let log_str = log.pretty()
+                            .replacen('\n', "- ", 1)  // Remove empty first line
+                            .replace('\n', "\n  ");  // Indent
+                        writeln!(output, "{}", log_str)?;
+                    }
+                },
+                // Break on Ctrl+C signal, to allow for closing JSON bracket
+                _ = ctrl_c() => {
```

Comment:
Could probably pass this signal as a generic cancel signal?

Reply:
I think for now this is OK—albeit a more generic one might be nicer for ppl doing more complex stuff?